File: Python/dm-git.py
import urllib.request
from re import compile
from json import loads


def build_dm():
    """
    Workaround that uses latest commits from https://github.com/DJScias/Discord-Datamining with GitHub's API
    (https://api.github.com/repos/DJScias/Discord-Datamining/commits)
    Limitations: Exclusive to canary, gets latest data only
    :return:
    """

    the_url = "https://api.github.com/repos/DJScias/Discord-Datamining/commits"

    headers = {
        "User-Agent": "*"
    }
    requ = (urllib.request.urlopen(urllib.request.Request(the_url, headers=headers)).read()).decode('utf-8')

    data = ''

    try:
        data = loads(requ)[0]["commit"]["message"]
    except():
        pass

    canary_reg = compile("[A-Za-z0-9]+ \(Canary build: [0-9]+\)")
    canary_bn_reg = compile('Canary build: [0-9]+')
    canary_h_reg = compile("[A-Za-z0-9]+")

    canary_data = str(canary_reg.findall(data))
    canary_bn = str(canary_bn_reg.findall(canary_data)[0]).split(":")[-1].replace(" ", "")
    canary_h = str(canary_h_reg.findall(canary_data)[0])
    canary_id = canary_h[0:7:1]

    return canary_bn, canary_h, canary_id

File: JavaScript/dm-git.js
const https = require('https'),
      http2 = require('http2'),
      { URL } = require('url'),
      dmu = new URL("https://api.github.com/repos/DJScias/Discord-Datamining/commits");

/**
 * Workaround that uses latest commits from https://github.com/DJScias/Discord-Datamining with GitHub's API
 * (https://api.github.com/repos/DJScias/Discord-Datamining/commits)
 * Limitations: Exclusive to canary, gets latest data only
 * @returns {Promise<Object>}
 */
const buildDM = async function() {
    let options = {
        hostname: dmu.hostname,
        path: dmu.pathname,
        method: http2.constants.HTTP2_METHOD_GET,
        headers: {
            [http2.constants.HTTP2_HEADER_USER_AGENT]: '*'
        }
    };

    return new Promise(res => {
        https.request(options, resp => {
            let buffer = [];
            resp.on('data', d => buffer.push(d));
            resp.on('end', () => {
                let data;

                try { data = JSON.parse(Buffer.concat(buffer).toString())[0]; }
                catch (e) { res(e); }

                let search = (/[A-Za-z0-9]+ \(Canary build: [0-9]+\)/).exec(data.commit.message)[0];
                const buildNumber = (/Canary build: [0-9]+/).exec(search)[0].split(":").slice(-1)[0].trim();
                const buildHash = (/[A-Za-z0-9]+/).exec(search)[0];
                const buildID = buildHash.substring(0, 7);

                /* Return scheme
                * {
                *   buildNumber: <number>,
                *   buildHash: <hash>,
                *   buildID: <id>
                * }
                */

                res({
                    buildNumber,
                    buildHash,
                    buildID
                });
            });
        }).end();
    });
};

module.exports = { buildDM };
